### ISSUE TITLE
Graphs time_length

### ISSUE THREAD

**Issue by 41south** [poblabs/weewx-belchertown#157]
Here's a strange one

With graphs time_length set to 90000 I get a time span on my Barometer that runs from 11:59 yesterday to 0929 today (when I changed the length). Should it not run back to 0830 yesterday as 90000 is around 25 hrs. I seem to be missing about 3 1/2 hours (210 minutes).
I was aiming for 24 hours which is 86400 when I noticed this anomaly. It's in other charts too.

```
[[chart4]]
        title = Barometer
        type = spline
        time_length = 90000
        [[[barometer]]]
            color = "#BECC00"
            yaxis_softmin = 990
            yaxis_softmax = 1010
            yaxis_tickinterval = 5
            name = Barometer (past 24 hrs)
            [[[[numberFormat]]]]
                decimals = 1
                decimalPoint = "."
                thousandsSep = ""
```
![Screen Shot 2019-06-06 at 9 41 04 AM](https://user-images.githubusercontent.com/1725001/58992547-409f2700-883f-11e9-8045-38c0ffbbb1c5.png)

**Comment by poblabs** [comment]
This is a weird one because of the way weewx gets the "first record" of the time span and the date displayed on the tooltip. I ended up tricking it with the day group to show the starting record, and on week/month/year it shows the last record to get the hour. There's a bit of a manual override on it to do that. 

So anything outside of this gets the first record in the group which is usually the `:59` record. I haven't thought of a good way to stop that though. Need to do more reading on the weewx code

You can read my thoughts on this exact problem within my own code here: https://github.com/poblabs/weewx-belchertown/blob/master/bin/user/belchertown.py#L1530-L1535

**Comment by 41south** [comment]
Ah hah. And, as I watch it it appears to be 'back filling' the time span to fill in the missing time - so it's now running from 11:59 to 09:49 - so by the time 11:59 rolls around again I'll have the full 24hrs?

**Comment by poblabs** [comment]
No, it'll always be on the :59. 

I have to re-acclimate myself to this since it's been a few months. The database lookups have to start with a start time, and and end time - aka the timespan. 

[Because of this line](https://github.com/poblabs/weewx-belchertown/blob/master/bin/user/belchertown.py#L1543) where I'm combining the output from the database to the timestamp of the query. It selects whether to match the output to the start timestamp or the end timestamp. 

On day charts the start will be a :59, so I force it to use the end timestamp which is the :00. But on the week charts the start time will be the right DAY, but the end time would be the next DAY. 

The timestamps are all figured out in this block of code. https://github.com/poblabs/weewx-belchertown/blob/master/bin/user/belchertown.py#L957-L995

So I'm pretty sure this whole thing has something to do with the weewx's way on creating those timespans. I'm not sure if I'm doing something wrong - or if it's something in weewx? 

For giggles, if you change your time_length to "today", everything should line up on their correct minutes.

**Comment by poblabs** [comment]
I can replicate it - in fact I've been aware of this the whole time just haven't known how to fix it. Since everyone's database may have different times, I can't really take the time and add a minute with confidence that'll be the fix. I think I need to read up on `weeplot.utilities.scaletime`

![image](https://user-images.githubusercontent.com/3484775/58996780-a6e86300-87c7-11e9-97db-15d9fb02bb65.png)

**Comment by 41south** [comment]
Oh, no, it's not the minute that I'm worried about it was the graph span. Initially it was less than the 25 hours but now as the day rolls on the span is 'filling in' as such. It still starts at 11:59 yesterday (thats fine) so it looks like the 'rolling' 24 hours may start once it gets past that 96000/25hr mark. I guess what's strange is that if I changed the graph at 0930 today why didn't it go back to 0830 yesterday (or 0829 if you like). I'll monitor it and let you know, but I suspect it's going to work itself out

**Comment by poblabs** [comment]
Hmm you're right. 25 hours ago for me should be 6:55pm, but my latest graph is 8:59pm. 

Maybe the `time_length = 90000` defined under the chart isn't rolling up and it's using the `time_length = today` from the graph group (in my scenario)? but still doesn't make sense since it should show from midnight - not 9pm yesterday. I'll have to dig into this deeper. Still think it's something with the scaletime utility from weewx. I might be using it wrong?

**Comment by 41south** [comment]
Yes correct, today should start at midnight, which is indeed working fine for the today graphs. I will leave it with you to contemplate :) It's almost like it could only retrospectively fill in so much data then it had to do a catchup from new data.

**Comment by 41south** [comment]
So ...... when the end (right hand side with new data) of the graph passed 11:59 the beginning jumped forward 3 hrs from 11:59 to 14:59. ???? Weird indeed

**Comment by poblabs** [comment]
Yeah, I'm trying to follow this one but I think I'm confusing myself quick. 

* Time now: `21:10`
* time_length: `90000` (25 hours)
* plotgen_ts: `1559783400` (plotgen_ts = self.gen_ts is the generation time). This is 21:10
* For the timespan: 
    * `plotgen_ts - int(time_length)`: - This is generation time, minus the 25 hours gives us `1559693400 `- this is yesterday at 20:10. This matches 25 hours ago. 
    * `weeplot.utilities.scaletime(plotgen_ts - int(time_length), plotgen_ts)` this is to get the timespan for 25 hours ago to right now. Yesterday 20:10 to today 21:10. 
        * Returns: `(1559707200, 1559793600, 10800)`
        * This means the start timestamp is `1559707200 `which is June 5 at midnight and the end timestamp is `1559793600 `which is June 6 midnight.
        * `10800 `is the increment. 

So the start time is wrong. It should be the `plotgen_ts - int(time_length)` result. I wonder if I'm using the wrong weewx formula. I think I got this from my 0.9 skin though - so I'm very lost. 😃

**Comment by poblabs** [comment]
Let's keep this simple and remove that `scaletime` utility. We'll make the start timestamp the current time minus the `time_length` (in this case 25 hours ago). The end timestamp is the current time. 

Colin you're still on 1.0rc9.1 so a few revisions behind. So if you want to try this it'll be by hand (if you don't want to upgrade). 

Open the `bin/user/belchertown.py` file and look for this line. Search for `minstamp` and you'll probably find it quick. https://github.com/poblabs/weewx-belchertown/blob/master/bin/user/belchertown.py#L995

Replace that entire line with this:

```
                    # Rolling timespans using seconds
                    minstamp = plotgen_ts - int(time_length) # Take the generation time and subtract the time_length to get our start time
                    maxstamp = plotgen_ts
```

Restart weewx and see how it goes. For me I am now seeing the chart 25 hours ago. 

Current time: 9:25pm. 25 hours ago would be 8:25pm:

![image](https://user-images.githubusercontent.com/3484775/59000457-9a203b00-87d8-11e9-8fcf-49da8f7dc844.png)

**Comment by biltmorelaker** [comment]
I've put this on my lab instance and it is working as it is for @poblabs. Will install on my production instance now.

**Comment by poblabs** [comment]
@biltmorelaker Awesome, thanks! I just pushed another update which should fix the `:59` in the tooltip time too. Give that a try and let me know

**Comment by biltmorelaker** [comment]
Yes! Seems to be just right. My graphs for "Day" are set to Last 25 Hours and are exactly that. I just pushed the changes to my production site https://wx.kutzenco.com. It's working there. Nice catch on the part of @41south and nice fix on the part of @poblabs!

**Comment by 41south** [comment]
Perfect, fixed here too. Thanks @poblabs! And thanks @biltmorelaker for the testing :)

I now have 12 hrs and 24hrs set and working

Good night to you both (still early afternoon here though) 😄

**Comment by poblabs** [comment]
Great! Glad it's working!

**Comment by poblabs** [comment]
This will be released in 1.0.1 on Saturday (my time), unless something pops up which delays it.
